File: tools/v2/order_packet_filters.py
"""Mirror packages/engine2/src/orderPacketFilters.ts — dedupe moves, then influence clip."""

from __future__ import annotations

from typing import List

from .influence_budget import filter_orders_by_influence_budget
from .state import Order


def dedupe_moves_one_per_force(orders: List[Order]) -> List[Order]:
    seen: set[str] = set()
    kept: List[Order] = []
    for o in orders:
        if o.kind == "move":
            fid = o.payload.get("force_id") or o.payload.get("forceId")
            if isinstance(fid, str):
                if fid in seen:
                    continue
                seen.add(fid)
        kept.append(o)
    return kept


def sanitize_player_orders(start_influence: int, orders: List[Order]) -> List[Order]:
    return filter_orders_by_influence_budget(
        start_influence, dedupe_moves_one_per_force(orders)
    )

File: packages/engine2/src/orderPacketFilters.ts
/**
 * Client-side order packet cleanup mirroring what we want humans and LLMs to respect
 * before calling `tick()`: duplicate moves for the same force are meaningless (only
 * the first can apply), then influence budgeting (see influenceBudget.ts).
 */

import { filterOrdersByInfluenceBudget } from "./influenceBudget.js";
import type { Order } from "./types.js";

/** Keep the first `move` for each `forceId` in packet order; drop later duplicates. */
export function dedupeMovesOnePerForce(orders: readonly Order[]): Order[] {
  const seen = new Set<string>();
  const out: Order[] = [];
  for (const o of orders) {
    if (o.kind === "move") {
      if (seen.has(o.forceId)) continue;
      seen.add(o.forceId);
    }
    out.push(o);
  }
  return out;
}

/**
 * Dedupe moves, then drop orders that would fail for insufficient Influence
 * (two-phase resolution in `tick()`).
 */
export function sanitizePlayerOrders(
  startInfluence: number,
  orders: readonly Order[],
): Order[] {
  return filterOrdersByInfluenceBudget(
    startInfluence,
    dedupeMovesOnePerForce(orders),
  );
}

/** True if `sanitizePlayerOrders` would remove at least one order. */
export function wouldClipOrders(
  startInfluence: number,
  orders: readonly Order[],
): boolean {
  return sanitizePlayerOrders(startInfluence, orders).length < orders.length;
}
